Docker Engineのバージョンが25.0以上だとイメージの解析ができない

## 実行環境
```
Client: Docker Engine - Community
 Version:    25.0.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.23.3
    Path:     /usr/local/lib/docker/cli-plugins/docker-compose
```

## テストのエラー
```
=== RUN   Test_checkPackageInformation
    check_pkg_test.go:13:
                Error Trace:    check_pkg_test.go:13
                Error:          Received unexpected error:
                                failed to parse image tar file: could not find image config
                Test:           Test_checkPackageInformation
--- FAIL: Test_checkPackageInformation (4.05s)
=== RUN   Test_outPackageVersionDiff
--- PASS: Test_outPackageVersionDiff (0.00s)
FAIL
FAIL    github.com/tklab-group/docker-image-disassembler/cli/checkpkg   4.069s
=== RUN   Test_getAptPkgInfoInImageFromDfile
    disassembler_test.go:11:
                Error Trace:    disassembler_test.go:11
                Error:          Received unexpected error:
                                failed to parse image tar file: could not find image config
                Test:           Test_getAptPkgInfoInImageFromDfile
--- FAIL: Test_getAptPkgInfoInImageFromDfile (4.17s)
```

## 調査

saveコマンドで得たmainfestファイルの持つデータデータの形式が変わっている

### 以前のmanifest.jsonの例
```json
[
    {
        "Config": "557db50c9ef1cd918b1ff78880c65f2ff6a450c646e324691a6fe7bc2aa3debc.json",
        "RepoTags": [
            "add-file:latest"
        ],
        "Layers": [
            "1e33021cb9e4fc8e989445d68584dd2386b15e902fc4e95fba64a5a917759466/layer.tar",
            "4a0c493ec850512a4f72774b55b08d61fe6a0ab80cda2165cab36f1e20b1429c/layer.tar",
            "9027e3575c9e8b582f792d5cd342c3cb0144a170413c98fd092432f782a6d952/layer.tar"
        ]
    }
]
```

### 25.0でのmanifest.jsonの例
```json
[
    {
        "Config": "blobs/sha256/35612d56a1ccfb899bb9e2fc4ff1a03cb41acbe2a083f5417bb496d8247d4d0b",
        "RepoTags": null,
        "Layers": [
            "blobs/sha256/138064343453a380fa50f129b36070ab5b4291a35862f75b43abc8217c008f06",
            "blobs/sha256/7d32f8afe43c636fbc64f55eb7fe966f669bb995edae858b5cc0e06ee87b5a84"
        ],
        "LayerSources": {
            "sha256:138064343453a380fa50f129b36070ab5b4291a35862f75b43abc8217c008f06": {
                "mediaType": "application/vnd.oci.image.layer.v1.tar",
                "size": 73699328,
                "digest": "sha256:138064343453a380fa50f129b36070ab5b4291a35862f75b43abc8217c008f06"
            },
            "sha256:7d32f8afe43c636fbc64f55eb7fe966f669bb995edae858b5cc0e06ee87b5a84": {
                "mediaType": "application/vnd.oci.image.layer.v1.tar",
                "size": 47312384,
                "digest": "sha256:7d32f8afe43c636fbc64f55eb7fe966f669bb995edae858b5cc0e06ee87b5a84"
            }
        }
    }
]
```

[Docker Engine 25.0 release notes | Docker Docs](https://docs.docker.com/engine/release-notes/25.0/)によれば、OCI互換にしたりするためのsaveで得られるtarファイルの形式を変更をしたとのこと

--- a/disassembler/image/image_archive.go
+++ b/disassembler/image/image_archive.go
@@ -2,13 +2,17 @@ package image
 
 import (
 	"archive/tar"
+	"bytes"
 	"compress/gzip"
+	"encoding/json"
+	"errors"
 	"fmt"
-	dockerimage "github.com/docker/docker/image"
-	"github.com/tklab-group/docker-image-disassembler/disassembler/filetree"
 	"io"
 	"path"
 	"strings"
+
+	dockerimage "github.com/docker/docker/image"
+	"github.com/tklab-group/docker-image-disassembler/disassembler/filetree"
 )
 
 type ImageArchive struct {
@@ -71,6 +75,46 @@ func NewImageArchive(tarFile io.Reader) (*ImageArchive, error) {
 
 				jsonFiles[name] = fileBuffer
 			}
+		} else if strings.HasPrefix(name, "blobs/") {
+			// For the OCI-compatible image format (used since Docker 25)
+
+			buffer := make([]byte, 1024)
+			n, err := io.ReadFull(tarReader, buffer)
+			if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
+				return img, err
+			}
+
+			// Only try reading a TAR if file is "big enough"
+			if n == cap(buffer) {
+				var unwrappedReader io.Reader
+				unwrappedReader, err = gzip.NewReader(io.MultiReader(bytes.NewReader(buffer[:n]), tarReader))
+				if err != nil {
+					// Not a gzipped entry
+					unwrappedReader = io.MultiReader(bytes.NewReader(buffer[:n]), tarReader)
+				}
+
+				// Try reading a TAR
+				layerReader := tar.NewReader(unwrappedReader)
+				tree, err := processLayerTar(name, layerReader)
+				if err == nil {
+					// add the layer to the image
+					img.LayerMap[tree.LayerName] = tree
+					continue
+				}
+			}
+
+			// Not a TAR (or smaller than our buffer), might be a JSON file
+			decoder := json.NewDecoder(bytes.NewReader(buffer[:n]))
+			token, err := decoder.Token()
+			if _, ok := token.(json.Delim); err == nil && ok {
+				// Looks like a JSON object (or array)
+				fileBuffer, err := io.ReadAll(io.MultiReader(bytes.NewReader(buffer[:n]), tarReader))
+				if err != nil {
+					return img, err
+				}
+				jsonFiles[name] = fileBuffer
+			}
+			// Ignore every other unknown file type
 		}
 	}
 

